Import mothur datatypes

shiltemann opened the pull request:
adds mothur datatypes to Galaxy

xRef: galaxyproject/tools-iuc/pull/449

There are some example files for each of the sniffers here: https://bioinf-galaxian.erasmusmc.nl/public/mothur/galaxy-sniffertest/ 

ping @IyadKandalaft @jj-umn @oxyko

IyadKandalaft committed:
initial files by IyadKandalaft

shiltemann committed:
add mothur prefix

shiltemann committed:
use 'with open' throughout

shiltemann committed:
minor changes to list creations

shiltemann committed:
remove sniffers for datatypes that were subclassed

shiltemann committed:
add mothur prefix to extension, update sniffers and sniffer order, subclass taxonomy datatypes

shiltemann committed:
remove phylip datatype (not specific to mothur)

shiltemann committed:
make sniffers for otu and distmatrices stricter

shiltemann committed:
remove two unused classes

bgruening commented on lib/galaxy/datatypes/converters/ref_to_seq_taxonomy_converter.xml:
```diff
@@ -0,0 +1,12 @@
+<tool id="CONVERTER_ref_to_seq_taxomony" name="Convert Ref taxonomy to Seq Taxonomy" version="1.0.0">
+  <description>converts 2 or 3 column sequence taxonomy file to a 2 column mothur taxonomy_outline format</description>
+  <command interpreter="python">ref_to_seq_taxonomy_converter.py "$input" "$output"</command>
```
`interpreter` is deprecated please use `python $__tool_directory__` instead.

bgruening commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,949 @@
+    file_ext = 'mothur.sabund'
+    def __init__(self, **kwd):
+        """
+        # http://www.mothur.org/wiki/Sabund_file
```
hash is not needed

bgruening commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,949 @@
+        Determines whether the file is a otu (operational taxonomic unit) format
+        """
+        try:
+            with open( filename ) as fh:
```
maybe you can use get_headers() here? https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/sniff.py#L192

bgruening commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,949 @@
+    def sniff( self, filename ):
+        """
+        Determines whether the file is a otu (operational taxonomic unit) format
+        """
```
You could also think about some doc-tests like this one: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/xml.py#L31

shiltemann replied:
ah, cool, will do, thanks :)

shiltemann committed:
stop using deprecated 'interpreter'

bgruening commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,949 @@
+    def __init__(self, **kwd):
+        Otu.__init__( self, **kwd )
+        # self.column_names[0] = ['label']
+        # self.column_names[1] = ['group']
```
you can remove obsolete stuff if you want

bgruening commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,949 @@
+            comment_lines = 0
+            ncols = 0
+            try:
+                with open( dataset.file_name ) as fh:
```
get_header()

bgruening commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,949 @@
+        The first line is column headings as of Mothur v 1.20
+        """
+        try:
+            with open( filename ) as fh:
```
get_header()

bgruening commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,949 @@
+
+class AlignReport(Tabular):
+    """
+QueryName	QueryLength	TemplateName	TemplateLength	SearchMethod	SearchScore	AlignmentMethod	QueryStart	QueryEnd	TemplateStart	TemplateEnd	PairwiseAlignmentLength	GapsInQuery	GapsInTemplate	LongestInsert	SimBtwnQuery&Template
```
indentation

bgruening commented:
@shiltemann this is great! I guess you need to fix many pep8 warnings and usage of get_header() is recommended as well as doc-tests.

Thanks a lot!

shiltemann commented:
@bgruening will do, thanks! :)

shiltemann committed:
stylefixes

shiltemann committed:
subclass mothur.align

shiltemann committed:
update sniffers and add doctests

shiltemann committed:
stylefixes and updates set_metas

yhoogstrate commented on lib/galaxy/datatypes/converters/ref_to_seq_taxonomy_converter.py:
```diff
@@ -0,0 +1,33 @@
+assert sys.version_info[:2] >= (2, 4)
+
+
+def stop_err(msg):
```
stop_err is not used elsewhere

yhoogstrate commented on lib/galaxy/datatypes/converters/ref_to_seq_taxonomy_converter.py:
```diff
@@ -0,0 +1,33 @@
+    outfile = open(sys.argv[2], 'w')
+    for i, line in enumerate(file(infile_name)):
+        line = line.rstrip()
+        if not line or line.startswith('#'):
```
What's your opinion on:
`if line and not line.startswith('#'):` and increase indentation of the following 3 lines

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+
+        headers = get_headers(dataset.file_name, sep='\t')
+        for line in headers:
+            try:
```
`if len(line) > 1` allows you to get rid of the `except` + `pass` construction

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+        count = 0
+        for line in headers:
+            if not line[0].startswith('@') and not line[0].startswith('#'):
+                if len(line) == 2 and re.match('forward|reverse', line[0]):
```
line[0] in ['forward','reverse'] ?

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+                if len(line) == 2 and re.match('forward|reverse', line[0]):
+                    count += 1
+                    continue
+                elif len(line) == 3 and re.match('barcode', line[0]):
```
regex necessary?

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+        """
+        headers = get_headers(filename, sep='\t', count=300)
+        count = 0
+        pat = '^([^ \t\n\r\x0c\x0b;]+([(]\\d+[)])?(;[^ \t\n\r\x0c\x0b;]+([(]\\d+[)])?)*(;)?)$'
```
Could you make a static class member of the compiled regex (https://docs.python.org/2/library/re.html#re.compile) ?

E.g at line 744:
`pat_prog = re.compile('^([^ \t\n\r\x0c\x0b;]+([(]\\d+[)])?(;[^ \t\n\r\x0c\x0b;]+([(]\\d+[)])?)*(;)?)$')`

and at line 783: `self.pat_prog.match(line[1])`

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+                    return False
+                if not re.match(pat, line[1]):
+                    return False
+                if not found_semicolons and str(line[1]).count(';') > 0:
```
count will count all occurances while find should stop after the first. what do you think of:

```
str(line[1]).find(';') > -1
```

I also think str() is unnecessary, otherwise the regex would fail in advance

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+        col_cnt = None
+        all_integers = True
+        for line in headers:
+            if count == 0:
```
How about:
`if count != 0:`
+ indent and remove `pass`

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+            flow_values = int(headers[0][0])
+            dataset.metadata.flow_values = flow_values
+        except:
+            pass
```
I think this pass should at least give a warning to the log file? You can take a look at the following example: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/binary.py#L512

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+
+    def make_html_table(self, dataset, skipchars=[]):
+        """Create HTML table, used for displaying peek"""
+        out = ['<table cellspacing="0" cellpadding="3">']
```
I think using arrays or lists is unnecessary and you could use the following instead:

```
 def make_html_table(self, dataset, skipchars=[]):
     """Create HTML table, used for displaying peek"""
     try:
         out = '<table cellspacing="0" cellpadding="3">'

         # Generate column header
         out += '<tr>'
         out += '<th>%d. Name</th>' % 1
         out += '<th>%d. Flows</th>' % 2
         for i in range(3, dataset.metadata.columns + 1):
             base = dataset.metadata.flow_order[(i + 1) % 4]
             out += '<th>%d. %d %s</th>' % (i - 2, base)
         out += '</tr>'
         out += self.make_html_peek_rows(dataset, skipchars=skipchars)
         out += '</table>'
     except Exception, exc:
         out = "Can't create peek %s" % str(exc)
     return out
```

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+
+            headers = get_headers(dataset.file_name, sep='\t', count=-1)
+            for line in headers:
+                if len(line) >= 2:
```
I was wondering whether for OTU files a comment line doesn't start with `@`. If so, I think it would be better to use `startswith('@')` which is consistent with the other code

yhoogstrate commented on lib/galaxy/datatypes/mothur.py:
```diff
@@ -0,0 +1,937 @@
+            dataset.metadata.columns = len(colnames)
+        if len(colnames) > 2:
+            dataset.metadata.groups = colnames[2:]
+            column_types = ['str']
```
If I understand it correctly, if `len(colnames)` is 2, no column_types will be set?

You could use:

```
dataset.metadata.column_types = ['str'] + (['int'] * ( len(headers[0]) -1))
```

outside of an if statement.

shiltemann committed:
add python optimisations

shiltemann committed:
allow for comment lines in Otu files

jmchilton committed:
Fixup exception handling in mothur datatypes.

jmchilton committed:
Add a sample sniff test for new Mothur datatype.

The existing sniffer tests with the mothur stuff tests the sniffers in isolation - this is more of an integration test that tests the sample datatype configuration for the newly added mothur stuff.

jmchilton commented:
I love these big multi-author PRs. These are the same datatypes JJ added to our MSI instance like 5 years ago I think - they've been on a long journey :smile: and I'm excited for them to be in Galaxy proper. I've opened a [PR](https://github.com/shiltemann/galaxy/pull/1) with some small tweaks and I'll happily merge these after that is merged downstream. Thanks a bunch @shiltemann et. al. - great work, great community effort.

shiltemann committed:
Merge pull request #1 from jmchilton/mothur_datatypes

Small tweaks to mothur dataytpes.

martenson commented:
yay! thanks a bunch @shiltemann @yhoogstrate @bgruening

bgruening commented:
This made my day! Thanks a bunch to all people involved. Good day for metagenomics!
